29 create checklist for tools in the toolkit eg aiverify (#38)
MinBZK/ai-validation: projects/tad/index.md

> # Transparency of algorithmic decision making

Reviewer: @uittenbroekrobbert shall we move this file to the docs directory? That way it will also appear on the site.

Reviewer 2: Ah, that's where it should be, yes!

(On a sidenote: @robbertbos we may want to have a short description for each of our projects/workstreams there as well?)

This document contains a checklist with requirements for tools we could use to help with the
transparency of algorithmic decision making.

The requirements are based on:

- **ISO 25010 standard:** This standard defines the eight quality characteristics and provides a

> framework for evaluating software quality ([source](https://www.iso.org/standard/35733.html)).

Reviewer 3: Besides the official ISO source, it might be an idea to also give a source where the actual 8 characteristics are specified.

- **Industry best practices:** This includes a broad range of recommendations and guidelines for IT tool
development and implementation.
- **Common IT tool requirements:** This information was gathered by analyzing various sources, such as
documentation from popular IT tools, user reviews, and articles from reputable tech publications
that discuss essential features and functionalities expected from different types of IT tools.
- **Internal discussion and common sense**: While above sources are already exhaustive, we also
used team discussions and our own knowledge.

## Overview of requirements

The requirements have been given a priority based on the [MoSCoW](https://en.wikipedia.org/wiki/MoSCoW_method)
scale to allow for tool comparison.

### Functionality

| Requirement                                                                                                                                            | Priority |
|--------------------------------------------------------------------------------------------------------------------------------------------------------|----------|

> | The tool allows users to perform technical tests on an algorithm or model.                                                                             | M        |

Reviewer 3: Technical tests can mean a lot of things. Maybe we can be more specific, for example technical tests to asses model performance, model bias, model fairness. Also, many of the technical tests must include a dataset, so a relevant dataset can also be input for technical tests.

| The tool allows users to choose which tests to perform.                                                                                                | M        |
| The tool allows users to fill out questionnaires to conduct impact assessments for AI. For example IAMA or ALTAI.                                      | M        |
| The tool can generate a human readable report.                                                                                                         | M        |
| The tools works with a standardized report format, that it can read, write, and update.                                                                | M        |
| The tool supports plugin functionality so additional tests can be added easily.                                                                        | S        |

> | The tool supports plugin functionality so additional tests can be added easily.                                                                        | S        |

Reviewer 3: ```suggestion
```

This item is a duplicate from the one above.

> <!-- markdownlint-disable MD013 -->

Reviewer: Maybe I missed the exact reason why we added this statement, but in the file `.markdownlint.yml` we set that MD013 should not be applied to tables. I think we could remove this markdownlint-disable statement here. What do you think?

Author: You are correct, this is no (longer) needed.

# Transparency of algorithmic decision making

This document contains a checklist with requirements for tools we could use to help with the transparency of algorithmic decision making.

The requirements are based on:

> - **ISO 25010 standard:** This standard defines the eight quality characteristics and provides a framework for evaluating software quality.

Reviewer: What do you think of adding references to relevant sources we have used?

For example:
```suggestion
- **ISO 25010 standard:** This standard defines the eight quality characteristics and provides a framework for evaluating software quality ([source](https://www.iso.org/standard/35733.html)).
```

- **Internal discussion and common sense**: While above sources are already exhaustive, we also used team discussions and our own knowledge.

The requirements have been given a priority based on the [MoSCoW](https://en.wikipedia.org/wiki/MoSCoW_method) scale to allow for tool comparison.

| ID   | Group                  | Requirement                                                  | Priority |
| ---- | ---------------------- | ------------------------------------------------------------ | -------- |
| 1... | Functionality          | The tool allows users to perform technical tests on an algorithm or model. | M        |
|      | Functionality          | The tool allows users to choose which tests to perform.      | M        |
|      | Functionality          | The tool supports plugin functionality so additional tests can be added easily. | S        |
|      | Functionality          | The tool can generate a human readable report.               | M        |

> |      | Functionality          | The tool can generate a machine readable report.             | S        |

Author: What is the aim for this? I can understand you'd want the option to save progress, update and generate, but I think that'd be seperate from (final) report option? Or are you aiming at a standarized format for reporting that could be imported, modified and saved? If it is the latter, maybe we should phrase it that way?

> |      | Functionality          | The tool can generate a machine readable report.             | S        |

Reviewer 2: What I meant is the standardized format for capturing technical test and process test (that we have a research ticket lined up for). I'm not sure that should be separate from a machine readable report?

> |      | Functionality          | The tool allows various formats for reports, like PDF, JSON, MD. | C        |

Reviewer 4: i would limit this to: the tool allows extension of report formats functionality.

Author: ```suggestion
|      | Functionality          | The tool allows extension of report formats functionality. | C        |
```

|      | Functionality          | The tool allows to create custom reports based on components. | S        |
|      | Functionality          | It is possible to add custom components for reports.         | S        |

> |      | Functionality          | The tool provides logging.                                   | S        |

Reviewer: Do we want to be more specific about logging? For example, logging different model versions, logging changes of impact assessments, logging technical tests for individual runs?

|      | Functionality          | The tool can be used on an isolated system without an internet connection. | S        |

> |      | Functionality          | The tool does not share any data or logging to anyone or anywhere. | C        |

Reviewer 4: not sharing any information anywhere, i find that a bit intense. What does it mean exactly? would prefer something like: this tool does not share any information to third parties

Author: I assume testing models can be 'sensitive', especially with the data being used. What this requirement should say is: no information, logging, data, anything, goes outside of the system, so not to AIVerify, not to Google Analytics.. not 2nd or 3rd party.. So, yes, the wording is a bit intense, though it is exactely what I meant. Maybe this wording is better?

```suggestion
|      | Functionality          | The tool operates with complete data privacy; it does not share any data or logging information. | C        |
```

|      | Functionality          | The tool can be integrated in a CI/CD flow.                  | C        |
|      | Functionality          | The tool can be offered as a (cloud) service where no local installation is required. | C        |

> |      | Reliability            | The tool can operate consistently and reliably.              | M        |

Reviewer 4: this one is a bit vague. Could you add what consistent and reliable means? When does it fail this check.

Author: Is this better?

```suggestion
|      | Reliability            | The tool operates consistently and reliably, meaning it delivers the same expected results every time you use it.             | M        |
```

> |      | Reliability            | The tool can recover automatically from common failures.     | M        |

Reviewer 4: i would make this a should

Author: ```suggestion
|      | Reliability            | The tool recovers automatically from common failures.     | S        |
```

> |      | Reliability            | The tool recovers from failures quickly and minimizes data loss. | S        |

Reviewer 4: When does this fail? can you make it more specific?

Author: ```suggestion
|      | Reliability            | This tool recovers from failures quickly, minimizing data loss, for example by automatically saving intermediate test progress results. | S        |
```

|      | Reliability            | The tool handles errors gracefully and informs users of any issues. | S        |
|      | Reliability            | The tool provides clear error messages and instructions for troubleshooting. | S        |
|      | Usability              | The tool possess a clean, intuitive, and visually appealing UI that follows industry standards. | S        |

> |      | Usability              | The tool offers keyboard shortcuts for efficient interaction. | C        |

Reviewer 4: This one is to much in my oppinion

Author: It most definately is a big 'could have' but I'm ok with it being a won't have.. it does however relate to the (only)  Accessibility issue, which requires accessibility features, in web applications (by default) keys like tab, enter etc have a (special) meaning so adding shortcuts to the wishlist is not that far fetched IMO.

Reviewer 2: We could actually instead refer to https://www.digitoegankelijk.nl/toegankelijkheid/en-301-549-en-wcag ?

Reviewer 5: I would agree with @anneschuth here if it would be tool that is client facing, but as for internal usage it wouldn't matter so much (if it is an CLI tool that we incorporate in our own UI).

Reviewer 2: I'm saying that the requirement should maybe focus on a11y in general (and then this would be the reference) instead of just shortcuts. 
And then a second question is how important this for the type of tool we're assessing.

Author: I have added the link to WCAG in the accessibility. I still do want to keep this requirement, even as a low 'could'. Usability is often overlooked and UX with keys can be a big plus, without having to be a big draw back if missing.

|      | Help & Documentation   | The tool provides comprehensive online help documentation with searchable functionalities. | S        |
|      | Help & Documentation   | The tool offers context-sensitive help within the application. | C        |
|      | Help & Documentation   | The online documentation includes video tutorials and training materials for ease of learning. | C        |
|      | Help & Documentation   | The project provides readily available customer support through various channels  (e.g., email, phone, online chat) to address user inquiries and troubleshoot issues. | C        |
|      | Performance Efficiency | The tool operates efficiently and minimize resource utilization. | M        |
|      | Performance Efficiency | The tool responds to user actions instantly.                 | M        |
|      | Performance Efficiency | Resource consumption of the tool itself (e.g., CPU, memory) does not exceed [Threshold] under normal load. Resource consumption of the tests can not be predicted. | C        |
|      | Performance Efficiency | The tool is scalable to accommodate increased user base and data volume. | S        |

> |      | Maintainability        | The project is open source.                                  | S        |

Reviewer 5: Although i would agree with S, maybe M would be better as it is one of our core principles to prefer open-source https://minbzk.github.io/ai-validation/Way-of-Working/Principles/

|      | Maintainability        | It is possible to contribute to the source.                  | S        |
|      | Maintainability        | The tool is easy to modify and maintain.                     | M        |
|      | Maintainability        | The tool adheres to industry coding standards and best practices to ensure code quality and maintainability. | M        |
|      | Maintainability        | The system is modular, allowing for easy modification of individual components. | S        |
|      | Maintainability        | Diagnostic tools are available to identify and troubleshoot issues. | S        |

> |      | Maintainability        | The code is written in a modern and common programming language. | M        |

Reviewer 5: Maybe drop the common? I like the modern as we wouldn't like tools in Fortran, but more specific languages could be Go which would also not be a game breaker it think (or would you still call this a common language?

Author: I think R is common enough... both common en modern are open to discussion ofcourse.. we could make it more specific (but that may make it too narrow):

The code is written in a common, widely adopted and supported and actively used and maintained programming language.

> |      | Maintainability        | The project provides version control for code changes and rollback capabilities. |          |

Reviewer 5: The priority was ommitted

```suggestion
|      | Maintainability        | The project provides version control for code changes and rollback capabilities. |  M        |
```

|      | Security               | The tool must protect data and system from unauthorized access, use, disclosure, disruption, modification, or destruction. | M        |
|      | Security               | The tool enforce authorization controls based on user roles and permissions, restricting access to sensitive data and functionalities. | C        |
|      | Security               | Data encryption is used for sensitive information at rest and in transit. | C        |
|      | Security               | Regular security audits and penetration testing are conducted. | S        |
|      | Security               | The project allows for regular security audits and penetration testing to identify vulnerabilities and ensure system integrity. | C        |
|      | Security               | The tool implements backup functionality to ensure data availability in case of incidents. | C        |
|      | Compatibility          | The tool is compatible with existing systems and infrastructure. | M        |
|      | Compatibility          | The tool supports industry-standard data formats and protocols. | M        |

> |      | Compatibility          | The tool operates seamlessly on supported operating systems and hardware platforms. | M        |

Author: I'm ok with an S. Though the essence of this requirement is: if the tool states it runs in X,Y,Z, it should run on X,Y,Z without any problems.

|      | Accessibility          | The tool is accessible to users with disabilities, following relevant accessibility standards (e.g., WCAG). | S        |
|      | Portability            | The tool support a range of operating systems (e.g., Windows, macOS, Linux) commonly used within an organization. | S        |
|      | Portability            | The tool minimizes dependencies on specific hardware or software configurations, promoting flexibility. | S        |
|      | Portability            | The tool offers a cloud-based deployment option or be compatible with cloud environments for scalability and accessibility. | S        |
|      | Portability            | The tool adheres to relevant cloud security standards and best practices. | S        |
|      | Deployment             | The tool has an easy and user-friendly installation and configuration process. | S        |
|      | Deployment             | The tool has on-premise or cloud-based deployment options to cater to different organizational needs and infrastructure. | S        |
|      | Legal & Compliance     | The tool is compliant with relevant legal and regulatory requirements. | S        |

> |      | Legal & Compliance     | The tool adheres to (local) data privacy regulations like GDPR and CCPA, ensuring the protection of user data. | S        |

Reviewer 2: If so, I'd argue the other way around (keep this one, drop the other one)

> |      | Legal & Compliance     | The tool adheres to (local) data privacy regulations like GDPR and CCPA, ensuring the protection of user data. | S        |

Author: This differse from just sharing data. It also tackles how to deal with storing data for example.. user sensitive data in an 'open' database (f.e. easy to guess login) or otherwise 'exposed'.. so I'd like to keep it.
